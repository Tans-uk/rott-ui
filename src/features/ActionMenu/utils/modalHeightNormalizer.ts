import {display} from '@utils'

export const modalHeightNormalizer = (
  dataLength: number,
  titleExist: boolean,
  subTitleExist: boolean,
  separatorTotalHeight: number,
  maxItem: number,
  itemHeight: number
) => {
  const showItemCount = dataLength > maxItem ? maxItem : dataLength
  const headerHeight = display.px(56)
  const cancelButtonHeight = display.px(56)
  const cancelButtonMarginTop = display.px(16)
  const listBottomPosition = display.px(24)

  return (
    ((showItemCount * itemHeight +
      (titleExist || subTitleExist ? headerHeight : 0) +
      cancelButtonHeight +
      cancelButtonMarginTop +
      listBottomPosition +
      separatorTotalHeight) *
      100) /
    display.setHeight(100)
  )
}
