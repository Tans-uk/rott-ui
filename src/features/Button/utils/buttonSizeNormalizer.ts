import {useDisplay} from '../../../hooks'
import type {Size} from '../../../models'

export const buttonSizeNormalizer = (buttonSize: Size | {width: string; height: string}) => {
  const {px, fontPixel} = useDisplay()

  switch (buttonSize) {
    case 'xs':
      return {
        width: px(85.5),
        height: px(36),
        fontSize: fontPixel(12),
        icon: px(18),
      }
    case 'sm':
      return {
        width: px(114),
        height: px(40),
        fontSize: fontPixel(14),
        icon: px(20),
      }
    case 'md':
      return {
        width: px(171),
        height: px(48),
        fontSize: fontPixel(14),
        icon: px(24),
      }
    case 'lg':
      return {
        width: px(228),
        height: px(56),
        fontSize: fontPixel(16),
        icon: px(24),
      }
    case 'xl':
    case 'xxl':
    case 'full':
    default:
      return {
        width: px(342),
        height: px(56),
        fontSize: fontPixel(16),
        icon: px(24),
      }
  }
}
