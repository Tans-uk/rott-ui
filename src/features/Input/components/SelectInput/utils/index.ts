export * from './listHeightNormalizer'
export * from './modalHeightPercentageNormalizer'
export * from './sortListBySearchPriority'
