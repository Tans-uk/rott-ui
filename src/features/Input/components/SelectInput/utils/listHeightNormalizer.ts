import {ITEM_HEIGHT, LIST_MAX_ITEM_COUNT, LIST_MIN_ITEM_COUNT, SEPARATOR_HEIGHT} from '../constants'

/**
 * Liste Yuksekligini Tanimlanan Min-Max Item Degerlerine Gore Hesaplar.
 * @param itemCount eleman sayisi
 * @returns liste yuksekligi
 */
export const listHeightNormalizer = (
  itemCount: number,
  showDescription: boolean,
  itemHeight: number = 56
) => {
  if (itemCount < LIST_MIN_ITEM_COUNT) {
    return (
      ITEM_HEIGHT(showDescription, itemHeight) * LIST_MIN_ITEM_COUNT +
      SEPARATOR_HEIGHT * (LIST_MIN_ITEM_COUNT + 1)
    )
  } else if (itemCount >= LIST_MAX_ITEM_COUNT) {
    return (
      ITEM_HEIGHT(showDescription, itemHeight) * LIST_MAX_ITEM_COUNT +
      SEPARATOR_HEIGHT * (LIST_MAX_ITEM_COUNT + 1)
    )
  } else
    return ITEM_HEIGHT(showDescription, itemHeight) * itemCount + SEPARATOR_HEIGHT * (itemCount + 1)
}
