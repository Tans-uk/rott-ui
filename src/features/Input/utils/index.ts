export * from './formatByDateMode'
export * from './useInputStyleNormalizer'
