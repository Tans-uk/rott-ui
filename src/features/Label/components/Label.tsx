import {forwardRef, type FC} from 'react'

import {StyleSheet, Text} from 'react-native'

import type {ThemeConfig} from '../../../models'
import type {LabelProps} from '../models'
import {useLabel} from '../styles'

export const Label: FC<LabelProps<ThemeConfig>> = forwardRef<Text, LabelProps<ThemeConfig>>(
  (
    {
      fontSize = 'md',
      variant = 'black',
      textCenter,
      fontWeight,
      fontFamily,
      letterSpacing = undefined,
      color,
      style,
      children,
      flex,
      ...props
    },
    ref
  ) => {
    const labelStyle = useLabel({
      textCenter,
      flex: flex ?? 1,
      fontSize,
      variant,
      fontWeight,
      fontFamily,
      letterSpacing,
      color,
      includeLatterSpacing: true,
      ...props,
    })

    return (
      <Text ref={ref} style={StyleSheet.flatten([labelStyle.defaultLabelStyle, style])} {...props}>
        {children}
      </Text>
    )
  }
)
