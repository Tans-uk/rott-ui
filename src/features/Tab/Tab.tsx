import {type FC} from 'react'

import {
  StyleSheet,
  Text,
  TouchableOpacity,
  type GestureResponderEvent,
  type LayoutChangeEvent,
  type TouchableOpacityProps,
} from 'react-native'

import {ThemeConfig, type CommonUiProps} from '../../models'
import {useTabStyle} from './styles'

interface TabProps<TTheme extends ThemeConfig>
  extends TouchableOpacityProps,
    CommonUiProps<TTheme> {
  title?: string
  isSelected?: boolean
  onPress?: (event: GestureResponderEvent) => void
  onLayout?: (event: LayoutChangeEvent) => void
}

export const Tab: FC<TabProps<ThemeConfig>> = ({
  title,
  isSelected,
  onPress,
  onLayout,
  ...props
}) => {
  const {container, textStyle} = useTabStyle(props)
  return (
    <TouchableOpacity
      style={container}
      onPress={onPress}
      onLayout={onLayout}
      aria-selected={isSelected}
      {...props}>
      <Text style={StyleSheet.flatten([textStyle])}>{title}</Text>
    </TouchableOpacity>
  )
}
