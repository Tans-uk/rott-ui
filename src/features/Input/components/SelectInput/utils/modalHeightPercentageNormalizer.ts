import {HEADER_HEIGHT, LIST_BOTTOM_MARGIN} from '../constants'
import {listHeightNormalizer} from './listHeightNormalizer'

import {display} from '@utils'

/**
 * Modal Yuksekligini Eleman Sayisina Gore Yuzde (%) Olarak Hesaplar
 * @param itemCount eleman sayisi
 * @returns Modal Yukseklik Yuzdesi
 */
export const modalHeightPercentageNormalizer = (
  itemCount: number,
  searchable: boolean,
  showDescription: boolean,
  itemHeight: number = 56
) => {
  const listHeight = listHeightNormalizer(itemCount, showDescription, itemHeight)

  const modalHeightPX = display.normalize(
    listHeight + HEADER_HEIGHT(searchable) + LIST_BOTTOM_MARGIN,
    'height'
  )

  return (modalHeightPX * 100) / display.percentage(100, 'height') + 1
}
