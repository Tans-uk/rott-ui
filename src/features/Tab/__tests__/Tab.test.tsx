import React from 'react'
import {commonUiTestExtension} from '../../../__tests__/utils/commonUiTestExtension'
import {fireEvent, render} from '../../../__tests__/utils/testUtils'
import {Label} from '../../Label'
import {Tab} from '../Tab'

const testId = {
  tabTestId: (name: string) => `${name}-test-id`,
}

describe('Tab -> Custom Component', () => {
  it('Tab snapshotı ile eşleşmeli', () => {
    const {tabTestId} = testId
    const rendered = render(
      <Tab testID={tabTestId('tab')} isSelected={false} onLayout={jest.fn()}>
        <Label>Test</Label>
      </Tab>
    )

    expect(rendered).toMatchSnapshot()
  })

  it('Tab onPress calismali', async () => {
    const {tabTestId} = testId
    const onPressMock = jest.fn()

    const {getByTestId} = render(
      <Tab testID={tabTestId('tab')} onPress={onPressMock}>
        <Label>Test-1</Label>
      </Tab>
    )

    let tabElement = getByTestId(tabTestId('tab'))
    fireEvent.press(tabElement)

    expect(onPressMock).toHaveBeenCalled()
  })

  it('Tab accessibility state degerlerinden selected olup olmadigi kontrol edilebilmeli', () => {
    const {tabTestId} = testId
    const returnsTrueMock = jest.fn(() => true)

    const {getByTestId} = render(
      <Tab testID={tabTestId('tab')} isSelected={returnsTrueMock()}>
        <Label>Test-1</Label>
      </Tab>
    )

    let tabElement = getByTestId(tabTestId('tab'))

    expect(tabElement).toHaveAccessibilityState({selected: true})
  })

  it('Tab selected degistigi zaman AccessibilityState degerlerinden selected degeri degismeli', () => {
    const {tabTestId} = testId
    let isSelectedMock = true

    const {getByTestId, rerender} = render(
      <Tab
        testID={tabTestId('tab')}
        isSelected={isSelectedMock}
        onPress={() => {
          isSelectedMock = !isSelectedMock
        }}>
        <Label>Test-1</Label>
      </Tab>
    )

    let tabElement = getByTestId(tabTestId('tab'))
    expect(tabElement).toHaveAccessibilityState({selected: true})

    fireEvent.press(tabElement)

    rerender(
      <Tab
        testID={tabTestId('tab')}
        isSelected={isSelectedMock}
        onPress={() => {
          isSelectedMock = !isSelectedMock
        }}>
        <Label>Test-1</Label>
      </Tab>
    )

    tabElement = getByTestId(tabTestId('tab'))

    expect(tabElement).toHaveAccessibilityState({selected: false})
  })
})

commonUiTestExtension(<Tab testID={testId.tabTestId('tab')} />, testId.tabTestId('tab'))
