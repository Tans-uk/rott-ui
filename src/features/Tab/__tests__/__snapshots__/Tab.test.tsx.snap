// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tab -> Custom Component Tab snapshotı ile eşleşmeli 1`] = `
<KeyboardProvider
  enabled={true}
>
  <View
    flex={1}
    pointerEvents="box-none"
    style={
      {
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": false,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={false}
      onClick={[Function]}
      onLayout={[MockFunction]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "opacity": 1,
          "width": 307,
        }
      }
      testID="tab-test-id"
    >
      <Text
        style={
          {
            "color": "rgba(234, 239, 240, 1)",
            "fontSize": 14,
          }
        }
      />
    </View>
    <View
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        [
          {
            "alignSelf": "center",
            "elevation": 999999,
            "flex": 0,
            "maxWidth": "100%",
            "position": "absolute",
            "zIndex": 999999,
          },
          {
            "flexDirection": "column-reverse",
            "justifyContent": "flex-start",
            "top": 30,
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          {
            "bottom": 0,
          }
        }
      >
        <RCTSafeAreaView />
      </View>
    </View>
    <View
      onLayout={[Function]}
      pointerEvents="box-none"
      style={
        [
          {
            "alignSelf": "center",
            "elevation": 999999,
            "flex": 0,
            "maxWidth": "100%",
            "position": "absolute",
            "zIndex": 999999,
          },
          {
            "bottom": 40,
            "flexDirection": "column",
            "justifyContent": "flex-end",
            "width": 750,
          },
        ]
      }
    >
      <View
        style={
          {
            "bottom": 0,
          }
        }
      >
        <RCTSafeAreaView />
      </View>
    </View>
  </View>
</KeyboardProvider>
`;
